### PR TITLE
fix(init): If init has errors, rollback changes

### PR DESCRIPTION
**Pull request by dustmop**
If init has errors, simply bailing with an error isn't good enough. It can leave our repo and working directory in unusable states. Instead, we create a cleanup function that will undo work done so far, calling any cleanup function that already exists.

Convert fsi_integration_test to new, simpler TestRunner methods.

### CODE REVIEW

File: fsi/init.go [hunk 1]
```diff
@@ -124,20 +163,37 @@ func (fsi *FSI) InitDataset(p InitParams) (name string, err error) {
...
 		if aComp != nil {
-			aComp.WriteTo(targetPath)
+			wroteFile, err := aComp.WriteTo(targetPath)
+			if err != nil {
+				return "", err
+			}
+			// If future steps fail, rollback the components that have been written
+			rollback = concatFunc(func() {
+				if wroteFile != "" {
+					log.Debugf("removing file \"%s\" during rollback", wroteFile)
```
[the remark this answers is not in the document]
Comment (dustmop) [reply fsi/init.go]: Even better, I just found %q which adds the double-quotes around a string.

File: fsi/fsi.go [hunk 2]
```diff
@@ -91,37 +91,56 @@ func (fsi *FSI) LinkedRefs(offset, limit int) ([]repo.DatasetRef, error) {
 	return refs, nil
 }
 
-// CreateLink connects a directory
-func (fsi *FSI) CreateLink(dirPath, refStr string) (string, error) {
+// CreateLink links a working directory to a dataset. Returns the reference alias, and a
+// rollback function if no error occurs
+func (fsi *FSI) CreateLink(dirPath, refStr string) (string, func(), error) {
```
[the remark this answers is not in the document]
Comment (dustmop) [reply fsi/fsi.go]: Unfortunately not. The zero value for func() is a nil pointer, which will segfault if called.

File: fsi/init.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (b5) [reply fsi/init.go]: Woah! Dope!

File: fsi/fsi.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (b5) [reply fsi/fsi.go]: fak, that's right. I should have tested this before commenting

**Pull request merged**